src/deploy/hosting/uploader.ts

```diff
@@ -0,0 +1,268 @@
+    this.public = options.public || this.cwd;
+    this.files = options.files;
+    this.fileCount = this.files.length;
+
+    this.cache = hashcache.load(this.projectRoot, this.hashcacheName());
+    this.cacheNew = new Map();
+
+    this.sizeMap = {};
+    this.hashMap = {};
+    this.pathMap = {};
+  }
+
+  hashcacheName(): string {
+    return Buffer.from(path.relative(this.projectRoot, this.public))
+      .toString("base64")
+      .replace(/=+$/, "");
+  }
+
+  public async start(): Promise<void> {
+    // short-circuit when there's zero files
+    if (this.files.length === 0) {
+      return;
+    }
+
+    for (const f of this.files) {
+      this.hashQueue.add(f);
+    }
+    this.hashQueue.close();
+    this.hashQueue.process();
+    this.hashQueue
+      .wait()
+      .then(this.queuePopulate.bind(this))
+      .then(() => {
+        hashcache.dump(this.projectRoot, this.hashcacheName(), this.cacheNew);
+        logger.debug("[hosting][hash queue][FINAL]", this.hashQueue.stats());
+        this.populateQueue.close();
+        return this.populateQueue.wait();
+      })
+      .then(() => {
+        logger.debug("[hosting][populate queue][FINAL]", this.populateQueue.stats());
+        logger.debug("[hosting] uploads queued:", this.uploadQueue.stats().total);
+        this.uploadQueue.close();
+      });
+
+    const fin = (err: unknown): void => {
+      logger.debug("[hosting][upload queue][FINAL]", this.uploadQueue.stats());
+      if (err) {
+        throw err;
+      }
+    };
+
+    return this.wait()
+      .then(fin)
+      .catch(fin);
+  }
+
+  async wait(): Promise<void> {
+    await Promise.all([this.hashQueue.wait(), this.populateQueue.wait(), this.uploadQueue.wait()]);
+  }
+
+  statusMessage(): string {
+    if (!this.hashQueue.finished) {
+      return progressMessage("hashing files", this.hashQueue.complete, this.fileCount);
+    } else if (!this.populateQueue.finished) {
+      return progressMessage(
+        "adding files to version",
+        this.populateQueue.complete * 1000,
+        this.fileCount
+      );
+    } else if (!this.uploadQueue.finished) {
+      return progressMessage(
+        "uploading new files",
+        this.uploadQueue.complete,
+        this.uploadQueue.stats().total
+      );
+    } else {
+      return "upload complete";
+    }
+  }
+
+  async hashHandler(filePath: string): Promise<void> {
+    const stats = fs.statSync(path.resolve(this.public, filePath));
+    const mtime = stats.mtime.getTime();
+    this.sizeMap[filePath] = stats.size;
+    const cached = this.cache[filePath];
+    if (cached && cached.mtime === mtime) {
+      this.cacheNew.set(filePath, cached);
+      this.addHash(filePath, cached.hash);
+      return;
+    }
+
+    const fstream = this.zipStream(filePath);
+    const hash = crypto.createHash("sha256");
+
+    fstream.pipe(hash);
+
+    return new Promise((resolve, reject) => {
+      fstream.on("end", () => {
+        const hashVal = hash.read().toString("hex");
```

Comment:
nit: Let's just `fstream.on("end", resolve);` and move these logic after awaiting the promise.

Reply:
seems reasonable enough of an adjustment. done.